### PR TITLE
enhance the handling of qubit_mapping

### PR DESCRIPTION
**Pull request by hay-k**
- Remove `IQMQubit` class and use `cirq.NameQubit` instead
- Automatically generate identity qubit mapping (using device qubit names) when not provided by the end user
- Add relevant tests

COMP-118

### CODE REVIEW

File: src/cirq_iqm/iqm_device.py [hunk 1]
```diff
@@ -67,6 +50,11 @@ class IQMDevice(devices.Device):
     Adds extra functionality on top of the basic :class:`cirq.Device` class for decomposing gates,
     optimizing circuits and mapping qubits.
     """
+    QUBIT_COUNT: int = None
+    """the number of qubits in the device"""
```
Comment (smite) [inline src/cirq_iqm/iqm_device.py:54]:
```suggestion
    """number of qubits on the device"""
```

File: src/cirq_iqm/iqm_device.py [hunk 2]
```diff
@@ -67,6 +50,11 @@ class IQMDevice(devices.Device):
...
+
+    QUBIT_NAME_PREFIX: str = 'QB'
+    """prefix for naming qubits based on their numerical index"""
```
Comment (smite) [inline src/cirq_iqm/iqm_device.py:57]:
```suggestion
    """prefix for qubit names, followed by their numerical index"""
```

File: src/cirq_iqm/iqm_device.py [hunk 3]
```diff
@@ -82,18 +70,12 @@ class IQMDevice(devices.Device):
...
-        expected = range(1, max(qubits) + 1)
-        missing = sorted(set(expected) - qubits)
-        if missing:
-            raise ValueError('{} connectivity map: the qubits {} are missing.'.format(self.__class__.__name__, missing))
+        self.qubits = tuple(cirq.NamedQubit.range(1, self.QUBIT_COUNT + 1, prefix=self.QUBIT_NAME_PREFIX))
 
-        self.qubits = tuple(IQMQubit(k) for k in expected)
+    @classmethod
+    def get_qubit_index(cls, qubit: cirq.NamedQubit):
+        """return the numeric index of the given qubit on device"""
```
Comment (smite) [inline src/cirq_iqm/iqm_device.py:77]:
```suggestion
        """The numeric index of the given qubit on the device."""
```

File: src/cirq_iqm/iqm_device.py [hunk 4]
```diff
@@ -82,18 +70,12 @@ class IQMDevice(devices.Device):
...
-
-        expected = range(1, max(qubits) + 1)
-        missing = sorted(set(expected) - qubits)
-        if missing:
-            raise ValueError('{} connectivity map: the qubits {} are missing.'.format(self.__class__.__name__, missing))
+        self.qubits = tuple(cirq.NamedQubit.range(1, self.QUBIT_COUNT + 1, prefix=self.QUBIT_NAME_PREFIX))
 
-        self.qubits = tuple(IQMQubit(k) for k in expected)
+    @classmethod
+    def get_qubit_index(cls, qubit: cirq.NamedQubit):
```
Comment (smite) [inline src/cirq_iqm/iqm_device.py:76]:
```suggestion
    def get_qubit_index(cls, qubit: cirq.NamedQubit) -> int:
```

File: src/cirq_iqm/iqm_device.py [hunk 5]
```diff
@@ -132,11 +114,16 @@ def map_circuit(self, circuit: cirq.Circuit, *, map_qubits: bool = True) -> cirq
         """
         # TODO permute circuit qubits to try to satisfy the device connectivity
         def map_qubit(qubit: cirq.ops.Qid) -> cirq.ops.Qid:
-            """NamedQubits are mapped to device qubits, other types of qubits are passed through as is."""
-            if isinstance(qubit, cirq.NamedQubit):
-                idx = int(re.search(r'\D*(\d+)?\D*', qubit.name).group(1))
-                return self.qubits[idx - 1]
-            return qubit
+            if not isinstance(qubit, cirq.NamedQubit):
+                raise ValueError('Only qubits of type cirq.NamedQubit are supported')
```
Comment (smite) [inline src/cirq_iqm/iqm_device.py:118]:
```suggestion
                raise ValueError('Only qubits of type cirq.NamedQubit are supported.')
```

File: src/cirq_iqm/iqm_device.py [hunk 6]
```diff
@@ -132,11 +114,16 @@ def map_circuit(self, circuit: cirq.Circuit, *, map_qubits: bool = True) -> cirq
...
+            if qubit in self.qubits:
+                return qubit
+
+            # map based on numeric identifier in the qubit name
+            idx = re.search(r'\D*(\d+)?\D*', qubit.name).group(1)
+            if idx is None:
+                raise ValueError('Qubit names do not contain numbering')
```
Comment (smite) [inline src/cirq_iqm/iqm_device.py:125]:
```suggestion
                raise ValueError('Qubit names must contain number.')
```

File: src/cirq_iqm/iqm_device.py (hunk 2 above)
Comment (hay-k) [reply src/cirq_iqm/iqm_device.py:57]: Maybe """prefix for qubit names, to be followed by their numerical index""" ?

**Pull request merged**